From cd1238744243e837f8b33f47f9cddff9d96de03c Mon Sep 17 00:00:00 2001
From: Clayton Craft <clayton@craftyguy.net>
Date: Mon, 15 Sep 2025 08:19:27 -0700
Subject: [PATCH] tree: support apple-nvme transport

Apple silicon support on Linux uses a separate platform driver called
'apple-nvme', and libnvme doesn't know about this transport. This fixes
a "No transport address for 'apple-nvme'" error when running on Apple
silicon, where libnvme was unable to enumerate nvme storage attached to
this platform.

$ nvme list
Node                  Generic               SN                   Model                                    Namespace  Usage                      Format           FW Rev
--------------------- --------------------- -------------------- ---------------------------------------- ---------- -------------------------- ---------------- --------
/dev/nvme0n1          /dev/ng0n1            xxxxxxxxxxxxxxxx     APPLE SSD AP2048Z                        0x1          2.00  TB /   2.00  TB      4 KiB +  0 B   532.140.
...

Signed-off-by: Clayton Craft <clayton@craftyguy.net>
---
 src/nvme/tree.c | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/src/nvme/tree.c b/src/nvme/tree.c
index d7970743..e4ac90a3 100644
--- a/src/nvme/tree.c
+++ b/src/nvme/tree.c
@@ -1417,7 +1417,8 @@ struct nvme_ctrl *nvme_create_ctrl(nvme_root_t r,
 		return NULL;
 	}
 	if (strncmp(transport, "loop", 4) &&
-	    strncmp(transport, "pcie", 4) && !traddr) {
+	    strncmp(transport, "pcie", 4) &&
+	    strncmp(transport, "apple-nvme", 10) && !traddr) {
 		nvme_msg(r, LOG_ERR, "No transport address for '%s'\n",
 			 transport);
 	       errno = EINVAL;
-- 
2.51.0

